src/Native/System.Native/pal_networking.cpp

```diff
@@ -379,7 +380,119 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR && (!HAVE_GETHOSTBYNAME_R && !HAVE_GETHOSTBYADDR_R)
```

Comment:
This structure is much cleaner. I will adapt it.
Regarding the assumption, what I have gathered is _r variants are GNUC specific. So if one is not present, then both will be absent. At least in the existing supported set of platforms, this assumption holds.